File: Nudge-Neurodivergent-task-tracker/HabitTracker.java
import java.time.LocalDate;
import java.util.HashMap;
import java.util.Map;
import java.util.List;
import java.util.ArrayList;

/**
 * Tracks habit chains and provides flexible habit building support
 */
public class HabitTracker {
    public static class Habit {
        private String name;
        private String description;
        private TaskCategory category;
        private Map<LocalDate, Boolean> completionHistory;
        private int targetFrequency; // times per week
        private boolean allowFlexibility;
        private LocalDate startDate;
        private int totalCompletions;
        
        public Habit(String name, String description, TaskCategory category, int targetFrequency) {
            this.name = name;
            this.description = description;
            this.category = category;
            this.targetFrequency = targetFrequency;
            this.completionHistory = new HashMap<>();
            this.allowFlexibility = true;
            this.startDate = LocalDate.now();
            this.totalCompletions = 0;
        }
        
        public void markCompleted(LocalDate date) {
            if (!completionHistory.getOrDefault(date, false)) {
                completionHistory.put(date, true);
                totalCompletions++;
            }
        }
        
        public void markSkipped(LocalDate date) {
            completionHistory.put(date, false);
        }
        
        public int getCurrentStreak() {
            int streak = 0;
            LocalDate checkDate = LocalDate.now();
            
            while (completionHistory.getOrDefault(checkDate, false)) {
                streak++;
                checkDate = checkDate.minusDays(1);
            }
            
            return streak;
        }
        
        public int getLongestStreak() {
            int maxStreak = 0;
            int currentStreak = 0;
            
            LocalDate checkDate = startDate;
            LocalDate today = LocalDate.now();
            
            while (!checkDate.isAfter(today)) {
                if (completionHistory.getOrDefault(checkDate, false)) {
                    currentStreak++;
                    maxStreak = Math.max(maxStreak, currentStreak);
                } else {
                    currentStreak = 0;
                }
                checkDate = checkDate.plusDays(1);
            }
            
            return maxStreak;
        }
        
        public double getWeeklyCompletionRate() {
            LocalDate weekStart = LocalDate.now().minusDays(6);
            int completions = 0;
            
            for (int i = 0; i < 7; i++) {
                LocalDate checkDate = weekStart.plusDays(i);
                if (completionHistory.getOrDefault(checkDate, false)) {
                    completions++;
                }
            }
            
            return targetFrequency > 0 ? (double) completions / targetFrequency : 0;
        }
        
        public String getProgressVisualization() {
            StringBuilder visual = new StringBuilder();
            LocalDate start = LocalDate.now().minusDays(6);
            
            for (int i = 0; i < 7; i++) {
                LocalDate checkDate = start.plusDays(i);
                if (completionHistory.getOrDefault(checkDate, false)) {
                    visual.append("🟢");
                } else if (completionHistory.containsKey(checkDate)) {
                    visual.append("🔴"); // Explicitly skipped
                } else {
                    visual.append("⚪"); // Not tracked yet
                }
            }
            
            return visual.toString();
        }
        
        public boolean isOnTrack() {
            double completionRate = getWeeklyCompletionRate();
            return allowFlexibility ? completionRate >= 0.7 : completionRate >= 1.0;
        }
        
        public String getEncouragementMessage() {
            int streak = getCurrentStreak();
            double rate = getWeeklyCompletionRate();
            
            if (streak >= 7) {
                return "🔥 You're on fire! " + streak + " days strong!";
            } else if (streak >= 3) {
                return "⭐ Great momentum! " + streak + " days in a row!";
            } else if (rate >= 0.8) {
                return "🎯 You're hitting your targets beautifully!";
            } else if (rate >= 0.5) {
                return "📈 Steady progress! Keep building that habit!";
            } else {
                return "🌱 Every day is a chance to restart. You've got this!";
            }
        }
        
        // Getters and setters
        public String getName() { return name; }
        public String getDescription() { return description; }
        public TaskCategory getCategory() { return category; }
        public int getTargetFrequency() { return targetFrequency; }
        public boolean isAllowFlexibility() { return allowFlexibility; }
        public LocalDate getStartDate() { return startDate; }
        public int getTotalCompletions() { return totalCompletions; }
        
        public void setAllowFlexibility(boolean allowFlexibility) {
            this.allowFlexibility = allowFlexibility;
        }
        
        public void setTargetFrequency(int targetFrequency) {
            this.targetFrequency = targetFrequency;
        }
        
        @Override
        public String toString() {
            return String.format("%s %s - %s (streak: %d)", 
                getProgressVisualization(), name, getEncouragementMessage(), getCurrentStreak());
        }
    }
    
    private List<Habit> habits;
    
    public HabitTracker() {
        this.habits = new ArrayList<>();
        createDefaultHabits();
    }
    
    private void createDefaultHabits() {
        // Common neurodivergent-friendly habits
        addHabit("Morning Medication", "Take prescribed medication", TaskCategory.MEDICATION, 7);
        addHabit("Hydration Check", "Drink a full glass of water", TaskCategory.HEALTH, 7);
        addHabit("5-Min Tidy", "Quick room/space organization", TaskCategory.CHORES, 5);
        addHabit("Gratitude Note", "Write one thing you're grateful for", TaskCategory.PERSONAL, 4);
        addHabit("Movement Break", "Any kind of physical movement", TaskCategory.HEALTH, 5);
    }
    
    public void addHabit(String name, String description, TaskCategory category, int targetFrequency) {
        habits.add(new Habit(name, description, category, targetFrequency));
    }
    
    public void removeHabit(String name) {
        habits.removeIf(habit -> habit.getName().equals(name));
    }
    
    public Habit findHabit(String name) {
        return habits.stream()
            .filter(habit -> habit.getName().equalsIgnoreCase(name))
            .findFirst()
            .orElse(null);
    }
    
    public void markHabitCompleted(String habitName, LocalDate date) {
        Habit habit = findHabit(habitName);
        if (habit != null) {
            habit.markCompleted(date);
        }
    }
    
    public void markHabitSkipped(String habitName, LocalDate date) {
        Habit habit = findHabit(habitName);
        if (habit != null) {
            habit.markSkipped(date);
        }
    }
    
    public List<Habit> getHabitsForToday() {
        return new ArrayList<>(habits);
    }
    
    public List<Habit> getHabitsByCategory(TaskCategory category) {
        return habits.stream()
            .filter(habit -> habit.getCategory() == category)
            .toList();
    }
    
    public void displayHabitDashboard() {
        System.out.println("\n" + "=".repeat(60));
        System.out.println("🌱 HABIT DASHBOARD");
        System.out.println("=".repeat(60));
        
        if (habits.isEmpty()) {
            System.out.println("💡 No habits tracked yet. Consider adding some small, manageable habits!");
            return;
        }
        
        System.out.println("📊 This Week's Progress (🟢 = done, 🔴 = skipped, ⚪ = not tracked)");
        System.out.println("-".repeat(60));
        
        for (Habit habit : habits) {
            System.out.println(habit.toString());
            
            double rate = habit.getWeeklyCompletionRate();
            String status = habit.isOnTrack() ? "✅ On track" : "⚠️ Needs attention";
            System.out.printf("   📈 Weekly rate: %.0f%% | %s\n", rate * 100, status);
            System.out.println();
        }
        
        displayHabitStats();
    }
    
    private void displayHabitStats() {
        int totalHabits = habits.size();
        long habitsOnTrack = habits.stream().mapToLong(habit -> habit.isOnTrack() ? 1 : 0).sum();
        int totalStreaks = habits.stream().mapToInt(Habit::getCurrentStreak).sum();
        int longestStreak = habits.stream().mapToInt(Habit::getLongestStreak).max().orElse(0);
        
        System.out.println("📊 OVERALL STATS:");
        System.out.printf("   🎯 Habits on track: %d/%d\n", habitsOnTrack, totalHabits);
        System.out.printf("   🔥 Active streaks: %d total days\n", totalStreaks);
        System.out.printf("   ⭐ Best streak ever: %d days\n", longestStreak);
        
        if (habitsOnTrack == totalHabits && totalHabits > 0) {
            System.out.println("   🎉 All habits on track - you're absolutely crushing it!");
        } else if (habitsOnTrack >= totalHabits * 0.7) {
            System.out.println("   💪 Most habits going strong - keep up the great work!");
        } else {
            System.out.println("   🌱 Room to grow - remember, progress over perfection!");
        }
    }
    
    public List<String> getTodaysHabitReminders() {
        List<String> reminders = new ArrayList<>();
        LocalDate today = LocalDate.now();
        
        for (Habit habit : habits) {
            if (!habit.completionHistory.getOrDefault(today, false)) {
                // Check if habit is due today based on frequency
                double weeklyRate = habit.getWeeklyCompletionRate();
                if (weeklyRate < (double) habit.getTargetFrequency() / 7) {
                    reminders.add("🌱 Habit reminder: " + habit.getName() + " - " + habit.getDescription());
                }
            }
        }
        
        return reminders;
    }
    
    public List<String> getHabitSuggestions(EnergyTracker.EnergyLevel energy) {
        List<String> suggestions = new ArrayList<>();
        
        switch (energy) {
            case VERY_LOW:
                suggestions.add("🌸 Focus only on essential habits (like medication)");
                suggestions.add("💙 Skip optional habits today - that's totally okay");
                break;
            case LOW:
                suggestions.add("🎯 Pick your top 2 most important habits");
                suggestions.add("⏰ Do them at the easiest time of day");
                break;
            case MODERATE:
                suggestions.add("📋 Good day for your regular habit routine");
                suggestions.add("🔄 Consider habit stacking (linking habits together)");
                break;
            case GOOD:
            case HIGH:
                suggestions.add("✨ Great energy for building new habit streaks!");
                suggestions.add("➕ Maybe add a small new habit to try");
                break;
        }
        
        return suggestions;
    }
    
    public String getMotivationalQuote() {
        String[] quotes = {
            "🌱 Small habits, big changes over time",
            "🔄 Consistency beats perfection every time", 
            "⭐ You're not building habits, you're becoming the person you want to be",
            "🎯 Focus on systems, not just goals",
            "💪 Every small action is a vote for the person you're becoming",
            "🌈 Progress is progress, no matter how small",
            "🔥 Habits are the compound interest of self-improvement"
        };
        
        return quotes[(int) (Math.random() * quotes.length)];
    }
    
    public List<Habit> getAllHabits() {
        return new ArrayList<>(habits);
    }
}

File: Nudge-Neurodivergent-task-tracker/habit_tracker.py
"""
Habit tracking system
"""

from datetime import date, timedelta
from typing import Dict, List, Set
from dataclasses import dataclass
import random
from task_manager import TaskCategory
from energy_tracker import EnergyLevel


@dataclass
class Habit:
    """Represents a habit to track"""
    name: str
    description: str
    category: TaskCategory
    target_frequency: int  # times per week
    completion_dates: Set[date]
    created_date: date
    
    def __post_init__(self):
        if not hasattr(self, 'completion_dates') or self.completion_dates is None:
            self.completion_dates = set()
    
    def mark_completed(self, completion_date: date = None):
        """Mark habit as completed for a specific date"""
        if completion_date is None:
            completion_date = date.today()
        self.completion_dates.add(completion_date)
    
    def is_completed_today(self) -> bool:
        """Check if habit was completed today"""
        return date.today() in self.completion_dates
    
    def get_weekly_progress(self) -> tuple[int, int]:
        """Get this week's progress (completed, target)"""
        today = date.today()
        week_start = today - timedelta(days=today.weekday())
        
        this_week_completions = sum(
            1 for d in self.completion_dates
            if week_start <= d <= today
        )
        
        return this_week_completions, self.target_frequency
    
    def get_current_streak(self) -> int:
        """Get current daily streak"""
        if not self.completion_dates:
            return 0
        
        streak = 0
        current_date = date.today()
        
        while current_date in self.completion_dates:
            streak += 1
            current_date -= timedelta(days=1)
        
        return streak
    
    def get_encouragement_message(self) -> str:
        """Get an encouraging message based on progress"""
        completed, target = self.get_weekly_progress()
        streak = self.get_current_streak()
        
        if completed >= target:
            messages = [
                "🎉 Weekly goal smashed! You're on fire!",
                "👑 Habit royalty! Goal achieved!",
                "⚡ Consistency champion! Target reached!"
            ]
        elif streak >= 3:
            messages = [
                f"🔥 {streak} day streak! Momentum building!",
                f"💪 {streak} days strong! Keep it rolling!",
                f"⭐ {streak} day consistency! You're amazing!"
            ]
        elif completed > 0:
            messages = [
                "🌱 Progress is progress! Every step counts!",
                "✨ Building that habit muscle! Great work!",
                "🎯 You're on the path! Keep going!"
            ]
        else:
            messages = [
                "🌟 Fresh start! Today's a perfect day to begin!",
                "💛 No pressure! Small steps lead to big changes!",
                "🌱 Every expert was once a beginner!"
            ]
        
        return random.choice(messages)


class HabitTracker:
    """Tracks habits and provides motivation"""
    
    def __init__(self):
        self.habits: List[Habit] = []
        self._create_starter_habits()
    
    def _create_starter_habits(self):
        """Create some starter habit suggestions"""
        # These are just examples - users can modify or remove them
        starter_habits = [
            Habit(
                name="Morning hydration",
                description="Drink a glass of water when you wake up",
                category=TaskCategory.SELF_CARE,
                target_frequency=7,
                completion_dates=set(),
                created_date=date.today()
            ),
            Habit(
                name="Daily medication",
                description="Take prescribed medication if applicable",
                category=TaskCategory.MEDICATION,
                target_frequency=7,
                completion_dates=set(),
                created_date=date.today()
            ),
            Habit(
                name="Gratitude moment",
                description="Notice one thing you're grateful for",
                category=TaskCategory.PERSONAL,
                target_frequency=5,
                completion_dates=set(),
                created_date=date.today()
            )
        ]
        self.habits = starter_habits
    
    def add_habit(self, name: str, description: str, category: TaskCategory, target_frequency: int):
        """Add a new habit"""
        habit = Habit(
            name=name,
            description=description,
            category=category,
            target_frequency=target_frequency,
            completion_dates=set(),
            created_date=date.today()
        )
        self.habits.append(habit)
        return habit
    
    def mark_habit_completed(self, habit_name: str, completion_date: date = None):
        """Mark a habit as completed"""
        for habit in self.habits:
            if habit.name.lower() == habit_name.lower():
                habit.mark_completed(completion_date)
                return True
        return False
    
    def get_todays_habit_reminders(self) -> List[str]:
        """Get gentle reminders for today's habits"""
        reminders = []
        incomplete_habits = [h for h in self.habits if not h.is_completed_today()]
        
        # Don't overwhelm - max 3 reminders
        for habit in incomplete_habits[:3]:
            completed, target = habit.get_weekly_progress()
            if completed < target:
                reminders.append(f"🌱 {habit.name} - {habit.description}")
        
        return reminders
    
    def display_habit_dashboard(self):
        """Display habit tracking dashboard"""
        print("\n🌱 HABIT TRACKER DASHBOARD")
        print("=" * 50)
        
        if not self.habits:
            print("No habits tracked yet. Consider adding some!")
            return
        
        for habit in self.habits:
            completed, target = habit.get_weekly_progress()
            streak = habit.get_current_streak()
            today_status = "✅" if habit.is_completed_today() else "⚪"
            
            print(f"\n{today_status} {habit.name}")
            print(f"   📊 This week: {completed}/{target}")
            print(f"   🔥 Streak: {streak} days")
            print(f"   📝 {habit.description}")
            
            # Progress bar
            progress = min(completed / target, 1.0) if target > 0 else 0
            bar_length = 20
            filled = int(progress * bar_length)
            bar = "█" * filled + "░" * (bar_length - filled)
            print(f"   [{bar}] {progress*100:.0f}%")
    
    def get_habit_suggestions(self, energy_level: EnergyLevel) -> List[str]:
        """Get habit suggestions based on energy level"""
        if energy_level.level_value <= 2:  # Low energy
            suggestions = [
                "🌸 Focus on micro-habits today (1-2 minutes)",
                "💧 Hydration habits are perfect for low energy",
                "🛁 Gentle self-care habits feel manageable",
                "📱 Use habit stacking with existing routines"
            ]
        elif energy_level.level_value >= 4:  # High energy
            suggestions = [
                "🚀 Great time to establish new habits",
                "💪 Consider increasing habit frequency goals",
                "🎯 Use this energy for habit planning",
                "⚡ Perfect for tackling avoided habits"
            ]
        else:  # Moderate energy
            suggestions = [
                "⚖️ Maintain your current habit rhythm",
                "🔄 Good time for habit consistency",
                "📋 Consider habit tracking review",
                "🌱 Steady progress is perfect progress"
            ]
        
        return suggestions
    
    def get_motivational_quote(self) -> str:
        """Get a motivational quote about habits"""
        quotes = [
            "We are what we repeatedly do. Excellence, then, is not an act, but a habit. - Aristotle",
            "Small habits, big changes. Your future self will thank you.",
            "Progress, not perfection. Every attempt builds the neural pathway.",
            "Habits are the compound interest of self-improvement.",
            "You don't have to be perfect, just consistent.",
            "The best time to plant a tree was 20 years ago. The second best time is now.",
            "Motivation gets you started. Habit keeps you going.",
            "Success is the sum of small efforts repeated day in and day out."
        ]
        return random.choice(quotes)
    
    def get_all_habits(self) -> List[Habit]:
        """Get all habits"""
        return self.habits.copy()
    
    def remove_habit(self, habit_name: str) -> bool:
        """Remove a habit"""
        for i, habit in enumerate(self.habits):
            if habit.name.lower() == habit_name.lower():
                self.habits.pop(i)
                return True
        return False